src/include/duckdb/common/types/string.hpp

```diff
@@ -0,0 +1,190 @@
+//===----------------------------------------------------------------------===//
+//                         DuckDB
+//
+// duckdb/common/types/string.hpp
+//
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include "duckdb/common/numeric_utils.hpp"
+#include "duckdb/common/typedefs.hpp"
+
+namespace duckdb {
+class String {
+public:
+	// Owning constructors
+	String(std::string str) // NOLINT: allow implicit conversion
+	    : owned_data(std::move(str)), data(owned_data.c_str()), size(owned_data.size()) {
+	}
+
+	String(const char *ptr, const size_t len) : owned_data(ptr, len), data(ptr), size(len) {
+	}
+
+	String(const char *ptr) // NOLINT: allow implicit conversion
+	    : String(ptr, strlen(ptr)) {
+	}
+
+	// Non-owning constructors
+	static String CreateView(const char *ptr, const size_t len) {
+		String str;
+		str.data = ptr;
+		str.size = len;
+		return str;
+	}
+
+	static String CreateView(const char *ptr) {
+		return CreateView(ptr, strlen(ptr));
+	}
+
+public:
+	bool operator==(const String &other) const {
+		const idx_t this_size = GetSize();
+		const idx_t other_size = other.GetSize();
+
+		if (this_size != other_size) {
+			return false;
+		}
+
+		const char *this_data = GetData();
+		const char *other_data = other.GetData();
+
+		for (idx_t i = 0; i < this_size; i++) {
+			if (this_data[i] != other_data[i]) {
+				return false;
+			}
+		}
+		return true;
+	}
+
+	bool operator!=(const String &other) const {
+		return !(*this == other);
+	}
+
+	bool operator<=(const String &other) const {
+		return *this < other || *this == other;
+	}
+
+	bool operator>=(const String &other) const {
+		return !(*this < other);
+	}
+
+	bool operator<(const String &other) const {
+		const char *this_data = GetData();
+		const char *other_data = other.GetData();
+		const idx_t this_size = GetSize();
+		const idx_t other_size = other.GetSize();
+
+		const idx_t length = MinValue<idx_t>(this_size, other_size);
+
+		for (idx_t i = 0; i < length; i++) {
+			if (this_data[i] < other_data[i]) {
+				return true;
+			}
+			if (this_data[i] > other_data[i]) {
+				return false;
+			}
+		}
+		return this_size < other_size;
+	}
+
+	bool operator>(const String &other) const {
+		return !(*this <= other);
+	}
+
+public:
+	idx_t GetSize() const {
+		return size;
+	}
+
+	const char *GetData() const {
+		return data;
+	}
+
+	char operator[](const idx_t pos) const {
+		D_ASSERT(pos < size);
+		return data[pos];
+	}
+
+	bool IsOwned() const {
+		return !owned_data.empty();
+	}
```

Comment:
I wonder if we want to add another field/boolean for this. Currently, we can also pass an empty string, which would then be treated as non-owning, even though it is. Not sure though, what do others think?

Reply:
I am not sure why we have the string, so can't say, isn't data pointer + size + an enum that decide how to implement delete is implemented enough?

Like it can be either:
- owned and delete has to be called on the data pointer
- owned but delete is implicit via arena
- non-owned